packages/contracts/src/LapuVault.sol

```diff
@@ -37,4 +52,71 @@ contract LapuVault is ERC4626 {
   function _convertToAssets(uint256 shares, Math.Rounding rounding) internal view override returns (uint256) {
     return shares;
   }
+
+  /**
+   * @dev emitting event Deposit(caller, receiver, assets, shares);
+   */
+  function deposit(uint256 assets, address receiver) public virtual override returns (uint256) {
+    uint256 maxAssets = maxDeposit(receiver);
+    if (assets > maxAssets) {
+      revert ERC4626ExceededMaxDeposit(receiver, assets, maxAssets);
+    }
+
+    uint256 shares = previewDeposit(assets);
+    _deposit(_msgSender(), receiver, assets, shares);
+
+    //deposit assets from this contract into DeFi pool
+    address poolAddress = poolAddressProvider.getPool();
+    IERC20(asset()).approve(poolAddress, assets);
+    IPool(poolAddress).supply(asset(), assets, address(this), 0);
+
+    return shares;
+  }
+
+  /**
+   * @dev emitting event Withdraw(caller, receiver, owner_, assets, shares)
+   */
+  function withdraw(uint256 assets, address receiver, address owner_) public virtual override returns (uint256) {
+    uint256 maxAssets = maxWithdraw(owner_);
+    if (assets > maxAssets) {
+      revert ERC4626ExceededMaxWithdraw(owner_, assets, maxAssets);
+    }
+
+    //withdraw assets from DeFi pool to this contract
+    address poolAddress = poolAddressProvider.getPool();
+    IPool(poolAddress).withdraw(asset(), assets, address(this));
+
+    uint256 shares = previewWithdraw(assets);
+    _withdraw(_msgSender(), receiver, owner_, assets, shares);
+
+    return shares;
+  }
+
+  /**
+   * @dev emitting event LAPUMintedFromDeFiYield(yield)
+   */
+  function mintLAPUAccordingToDeFiYield() external onlyOwner returns (uint256) {
+    uint256 currentLAPUTotalShare = totalSupply();
+    uint256 currentATokenBalance = aToken.balanceOf(address(this));
+    if (currentATokenBalance > currentLAPUTotalShare) {
+      uint256 yield = currentATokenBalance - currentLAPUTotalShare;
+      //mint the yield as new LAPU shares owned by this contract, such that they can be distributed to players as rewards at a later time
+      _mint(address(this), yield);
+      emit LAPUMintedFromDeFiYield(yield);
+      return yield;
+    } else {
+      return 0;
+    }
+  }
```

Comment:
If I understand correctly this supports only a single yield source at the moment (aToken)?

Reply:
that's correct